hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -141,7 +141,12 @@ def run_state_machine(factory, data):
                     machine._initialize_rules_to_run.remove(rule)
                 else:
                     rule, data = cd.draw(machine._rules_strategy)
-                data_to_print = dict(data)
+
+                # Pretty-print the values this rule was called with *before* calling
+                # _add_result_to_targets, to avoid printing arguments which are also
+                # a return value using the variable name they are assigned to.
+                # See https://github.com/HypothesisWorks/hypothesis/issues/2341
+                data_to_print = {k: machine._pretty_print(v) for k, v in data.items()}
```

Comment:
This ends up calling the pretty printing logic even when we're not printing, which is potentially fairly expensive because it runs on every single test case execution now. 

Would it maybe make more sense to refactor the code a bit so that `add_result_to_targets`, or at least the bit which updates the pretty printer, is called after printing?

Reply:
I tried that before, and couldn't find a structure that fixed the bug and otherwise preserved our existing behaviour.

Instead, I've just put an `if print_steps:` in front of the calculation - which means we're only shifting *when* we do the work, not how much.